File: phases/16-software-engineering-and-architecture/11-cqrs-and-event-sourcing/code/main.ts
// CQRS and Event Sourcing — TypeScript Implementation
// Phase 16 — Software Engineering & Architecture

// ─── Domain Events ───────────────────────────────────────────────────────────
// Events are immutable facts representing things that have happened in the domain.
// They are the source of truth — state is always derived by replaying events.

type AccountEvent =
  | { type: "AccountOpened"; accountId: string; initialDeposit: number; owner: string }
  | { type: "MoneyDeposited"; accountId: string; amount: number; description: string }
  | { type: "MoneyWithdrawn"; accountId: string; amount: number; description: string }
  | { type: "OverdraftLimitSet"; accountId: string; limit: number }
  | { type: "AccountClosed"; accountId: string };

// ─── Commands ────────────────────────────────────────────────────────────────
// Commands express intent to change state. They can be rejected if business rules
// are violated. Commands are processed by the command handler (write side).

type AccountCommand =
  | { type: "OpenAccount"; accountId: string; initialDeposit: number; owner: string }
  | { type: "Deposit"; accountId: string; amount: number; description: string }
  | { type: "Withdraw"; accountId: string; amount: number; description: string }
  | { type: "SetOverdraftLimit"; accountId: string; limit: number }
  | { type: "CloseAccount"; accountId: string };

// ─── Domain Errors ───────────────────────────────────────────────────────────

class DomainError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "DomainError";
  }
}

class AccountAlreadyOpenError extends DomainError {
  constructor(id: string) { super(`Account ${id} is already open`); }
}

class AccountNotFoundError extends DomainError {
  constructor(id: string) { super(`Account ${id} not found`); }
}

class AccountClosedError extends DomainError {
  constructor(id: string) { super(`Account ${id} is closed`); }
}

class InsufficientFundsError extends DomainError {
  constructor(balance: number, requested: number, overdraftLimit: number) {
    super(`Insufficient funds: balance=${balance}, requested=${requested}, overdraft_limit=${overdraftLimit}`);
  }
}

class InvalidAmountError extends DomainError {
  constructor(amount: number) { super(`Invalid amount: ${amount}`); }
}

class CannotDepositToClosedError extends DomainError {
  constructor(id: string) { super(`Cannot deposit to closed account ${id}`); }
}

class CannotWithdrawFromClosedError extends DomainError {
  constructor(id: string) { super(`Cannot withdraw from closed account ${id}`); }
}

// ─── Aggregate State ─────────────────────────────────────────────────────────
// The aggregate rebuilds its state by applying events in order.
// This is the core of event sourcing: state = events.reduce(apply, initialState).

interface AccountState {
  accountId: string;
  owner: string;
  balance: number;
  overdraftLimit: number;
  isOpen: boolean;
}

const initialState = (accountId: string): AccountState => ({
  accountId,
  owner: "",
  balance: 0,
  overdraftLimit: 0,
  isOpen: false,
});

function applyEvent(state: AccountState, event: AccountEvent): AccountState {
  switch (event.type) {
    case "AccountOpened":
      return {
        ...state,
        accountId: event.accountId,
        owner: event.owner,
        balance: event.initialDeposit,
        isOpen: true,
      };
    case "MoneyDeposited":
      return { ...state, balance: state.balance + event.amount };
    case "MoneyWithdrawn":
      return { ...state, balance: state.balance - event.amount };
    case "OverdraftLimitSet":
      return { ...state, overdraftLimit: event.limit };
    case "AccountClosed":
      return { ...state, isOpen: false };
  }
}

function rebuildState(accountId: string, events: AccountEvent[]): AccountState {
  return events.reduce(applyEvent, initialState(accountId));
}

// ─── Event Store ─────────────────────────────────────────────────────────────
// The event store is append-only. Events are never modified or deleted.
// In production: EventStoreDB, Kafka topics, or an append-only database table.

class EventStore {
  private streams: Map<string, AccountEvent[]> = new Map();

  append(accountId: string, events: AccountEvent[]): void {
    const stream = this.streams.get(accountId) ?? [];
    this.streams.set(accountId, [...stream, ...events]);
  }

  load(accountId: string): AccountEvent[] {
    return this.streams.get(accountId) ?? [];
  }

  loadFromVersion(accountId: string, fromVersion: number): AccountEvent[] {
    const stream = this.streams.get(accountId) ?? [];
    return stream.slice(fromVersion);
  }

  version(accountId: string): number {
    return this.streams.get(accountId)?.length ?? 0;
  }
}

// ─── Command Handler (Write Side) ───────────────────────────────────────────
// The command handler validates business rules and emits events.
// It never returns data — only events or errors. This is CQRS's write side.

class CommandHandler {
  constructor(private store: EventStore) {}

  handle(command: AccountCommand): AccountEvent[] {
    switch (command.type) {
      case "OpenAccount":
        return this.openAccount(command);
      case "Deposit":
        return this.deposit(command);
      case "Withdraw":
        return this.withdraw(command);
      case "SetOverdraftLimit":
        return this.setOverdraftLimit(command);
      case "CloseAccount":
        return this.closeAccount(command);
    }
  }

  private openAccount(cmd: AccountCommand & { type: "OpenAccount" }): AccountEvent[] {
    if (cmd.initialDeposit < 0) {
      throw new InvalidAmountError(cmd.initialDeposit);
    }
    const existing = this.store.load(cmd.accountId);
    if (existing.length > 0) {
      throw new AccountAlreadyOpenError(cmd.accountId);
    }
    return [{
      type: "AccountOpened",
      accountId: cmd.accountId,
      initialDeposit: cmd.initialDeposit,
      owner: cmd.owner,
    }];
  }

  private deposit(cmd: AccountCommand & { type: "Deposit" }): AccountEvent[] {
    if (cmd.amount <= 0) {
      throw new InvalidAmountError(cmd.amount);
    }
    const events = this.store.load(cmd.accountId);
    if (events.length === 0) {
      throw new AccountNotFoundError(cmd.accountId);
    }
    const state = rebuildState(cmd.accountId, events);
    if (!state.isOpen) {
      throw new CannotDepositToClosedError(cmd.accountId);
    }
    return [{
      type: "MoneyDeposited",
      accountId: cmd.accountId,
      amount: cmd.amount,
      description: cmd.description,
    }];
  }

  private withdraw(cmd: AccountCommand & { type: "Withdraw" }): AccountEvent[] {
    if (cmd.amount <= 0) {
      throw new InvalidAmountError(cmd.amount);
    }
    const events = this.store.load(cmd.accountId);
    if (events.length === 0) {
      throw new AccountNotFoundError(cmd.accountId);
    }
    const state = rebuildState(cmd.accountId, events);
    if (!state.isOpen) {
      throw new CannotWithdrawFromClosedError(cmd.accountId);
    }
    const available = state.balance + state.overdraftLimit;
    if (available < cmd.amount) {
      throw new InsufficientFundsError(state.balance, cmd.amount, state.overdraftLimit);
    }
    return [{
      type: "MoneyWithdrawn",
      accountId: cmd.accountId,
      amount: cmd.amount,
      description: cmd.description,
    }];
  }

  private setOverdraftLimit(cmd: AccountCommand & { type: "SetOverdraftLimit" }): AccountEvent[] {
    if (cmd.limit < 0) {
      throw new InvalidAmountError(cmd.limit);
    }
    const events = this.store.load(cmd.accountId);
    if (events.length === 0) {
      throw new AccountNotFoundError(cmd.accountId);
    }
    const state = rebuildState(cmd.accountId, events);
    if (!state.isOpen) {
      throw new AccountClosedError(cmd.accountId);
    }
    return [{
      type: "OverdraftLimitSet",
      accountId: cmd.accountId,
      limit: cmd.limit,
    }];
  }

  private closeAccount(cmd: AccountCommand & { type: "CloseAccount" }): AccountEvent[] {
    const events = this.store.load(cmd.accountId);
    if (events.length === 0) {
      throw new AccountNotFoundError(cmd.accountId);
    }
    const state = rebuildState(cmd.accountId, events);
    if (!state.isOpen) {
      throw new AccountClosedError(cmd.accountId);
    }
    return [{
      type: "AccountClosed",
      accountId: cmd.accountId,
    }];
  }
}

// ─── Projections (Read Model) ─────────────────────────────────────────────────
// Projections consume events and build read-optimized views. This is the
// query side of CQRS. Multiple projections can consume the same event stream.

interface BalanceView {
  accountId: string;
  balance: number;
  overdraftLimit: number;
  available: number;
  isOpen: boolean;
}

class BalanceProjection {
  static project(accountId: string, events: AccountEvent[]): BalanceView {
    const state = rebuildState(accountId, events);
    return {
      accountId: state.accountId,
      balance: state.balance,
      overdraftLimit: state.overdraftLimit,
      available: state.balance + state.overdraftLimit,
      isOpen: state.isOpen,
    };
  }
}

interface TransactionEntry {
  eventType: string;
  amount: number;
  description: string;
}

class HistoryProjection {
  static project(events: AccountEvent[]): TransactionEntry[] {
    return events.map((event): TransactionEntry => {
      switch (event.type) {
        case "AccountOpened":
          return {
            eventType: "OPENED",
            amount: event.initialDeposit,
            description: `Account opened by ${event.owner}`,
          };
        case "MoneyDeposited":
          return {
            eventType: "DEPOSIT",
            amount: event.amount,
            description: event.description,
          };
        case "MoneyWithdrawn":
          return {
            eventType: "WITHDRAWAL",
            amount: -event.amount,
            description: event.description,
          };
        case "OverdraftLimitSet":
          return {
            eventType: "OVERDRAFT_SET",
            amount: event.limit,
            description: `Overdraft limit set to ${event.limit}`,
          };
        case "AccountClosed":
          return {
            eventType: "CLOSED",
            amount: 0,
            description: "Account closed",
          };
      }
    });
  }
}

// ─── Snapshots ───────────────────────────────────────────────────────────────
// Snapshots save aggregate state at a point in time to avoid replaying
// all events on every command. This is essential for aggregates with many events.

const SNAPSHOT_INTERVAL = 3;

interface Snapshot {
  state: AccountState;
  version: number;
}

class SnapshotStore {
  private snapshots: Map<string, Snapshot> = new Map();

  save(accountId: string, state: AccountState, version: number): void {
    this.snapshots.set(accountId, { state, version });
  }

  load(accountId: string): Snapshot | undefined {
    return this.snapshots.get(accountId);
  }
}

function loadWithSnapshot(
  store: EventStore,
  snapshotStore: SnapshotStore,
  accountId: string,
): AccountState {
  const snapshot = snapshotStore.load(accountId);
  if (snapshot) {
    const remaining = store.loadFromVersion(accountId, snapshot.version);
    return remaining.reduce(applyEvent, snapshot.state);
  }
  const events = store.load(accountId);
  return rebuildState(accountId, events);
}

function maybeSaveSnapshot(
  snapshotStore: SnapshotStore,
  store: EventStore,
  accountId: string,
): void {
  const version = store.version(accountId);
  if (version > 0 && version % SNAPSHOT_INTERVAL === 0) {
    const events = store.load(accountId);
    const state = rebuildState(accountId, events);
    snapshotStore.save(accountId, state, version);
  }
}

// ─── Query Service (Read Side) ────────────────────────────────────────────────
// The query service reads from projections only. It never touches the event
// store directly for queries — that's CQRS. Commands go through the command
// handler (write side); queries go through projections (read side).

class QueryService {
  constructor(private store: EventStore) {}

  getBalance(accountId: string): BalanceView | null {
    const events = this.store.load(accountId);
    if (events.length === 0) return null;
    return BalanceProjection.project(accountId, events);
  }

  getHistory(accountId: string): TransactionEntry[] | null {
    const events = this.store.load(accountId);
    if (events.length === 0) return null;
    return HistoryProjection.project(events);
  }
}

// ─── Demo ─────────────────────────────────────────────────────────────────────

function main(): void {
  const store = new EventStore();
  const snapshots = new SnapshotStore();
  const handler = new CommandHandler(store);
  const accountId = "acc-001";

  console.log("=== CQRS + Event Sourcing: Bank Account Demo ===\n");

  // Open account
  const openEvents = handler.handle({
    type: "OpenAccount", accountId, initialDeposit: 100, owner: "Alice",
  });
  store.append(accountId, openEvents);
  maybeSaveSnapshot(snapshots, store, accountId);

  // Deposit
  const depositEvents = handler.handle({
    type: "Deposit", accountId, amount: 50, description: "Salary",
  });
  store.append(accountId, depositEvents);
  maybeSaveSnapshot(snapshots, store, accountId);

  // Set overdraft limit
  const overdraftEvents = handler.handle({
    type: "SetOverdraftLimit", accountId, limit: 200,
  });
  store.append(accountId, overdraftEvents);
  maybeSaveSnapshot(snapshots, store, accountId);

  // Withdraw within balance
  const withdrawEvents = handler.handle({
    type: "Withdraw", accountId, amount: 80, description: "Groceries",
  });
  store.append(accountId, withdrawEvents);
  maybeSaveSnapshot(snapshots, store, accountId);

  // Withdraw into overdraft
  const overdraftWithdrawEvents = handler.handle({
    type: "Withdraw", accountId, amount: 200, description: "Emergency repair",
  });
  store.append(accountId, overdraftWithdrawEvents);
  maybeSaveSnapshot(snapshots, store, accountId);

  // Try to exceed overdraft — should fail
  try {
    handler.handle({
      type: "Withdraw", accountId, amount: 100, description: "Should fail",
    });
  } catch (err) {
    console.log(`Withdrawal exceeding overdraft: ${(err as DomainError).message}`);
  }

  // Query: current balance via projection (READ SIDE)
  const queryService = new QueryService(store);
  const balanceView = queryService.getBalance(accountId)!;
  console.log("\n--- Balance View (Read Model) ---");
  console.log(`Account: ${balanceView.accountId}`);
  console.log(`Balance: $${balanceView.balance.toFixed(2)}`);
  console.log(`Overdraft Limit: $${balanceView.overdraftLimit.toFixed(2)}`);
  console.log(`Available: $${balanceView.available.toFixed(2)}`);
  console.log(`Is Open: ${balanceView.isOpen}`);

  // Query: transaction history via projection (READ SIDE)
  const history = queryService.getHistory(accountId)!;
  console.log("\n--- Transaction History (Read Model) ---");
  for (const entry of history) {
    console.log(`${entry.eventType.padEnd(15)} ${entry.amount.toFixed(2).padStart(8)}  ${entry.description}`);
  }

  // Verify snapshot replay matches full replay
  const fullState = rebuildState(accountId, store.load(accountId));
  const snapshotState = loadWithSnapshot(store, snapshots, accountId);
  console.log("\n--- Snapshot Verification ---");
  console.log(`Full replay balance: $${fullState.balance.toFixed(2)}`);
  console.log(`Snapshot replay balance: $${snapshotState.balance.toFixed(2)}`);
  console.assert(Math.abs(fullState.balance - snapshotState.balance) < Number.EPSILON,
    "Snapshot replay must match full replay");
  console.log("Snapshot replay matches full replay: OK");

  // Show the event store
  console.log("\n--- Event Store (Immutable Log) ---");
  const allEvents = store.load(accountId);
  allEvents.forEach((event, i) => {
    console.log(`v${i + 1}: ${JSON.stringify(event)}`);
  });

  // Close account
  const closeEvents = handler.handle({ type: "CloseAccount", accountId });
  store.append(accountId, closeEvents);

  // Try deposit on closed account — should fail
  try {
    handler.handle({
      type: "Deposit", accountId, amount: 10, description: "Should fail",
    });
  } catch (err) {
    console.log(`\nDeposit to closed account: ${(err as DomainError).message}`);
  }

  console.log("\n=== All business rules enforced. CQRS + ES demo complete. ===");
}

main();

File: phases/16-software-engineering-and-architecture/11-cqrs-and-event-sourcing/code/main.rs
use std::collections::HashMap;

// ─── Domain Events ───────────────────────────────────────────────────────────
// Events are immutable facts. Once something has happened, it cannot be undone.
// The event store records these facts in order, and we rebuild state by replaying.

#[derive(Debug, Clone, PartialEq)]
enum AccountEvent {
    AccountOpened { account_id: String, initial_deposit: f64, owner: String },
    MoneyDeposited { account_id: String, amount: f64, description: String },
    MoneyWithdrawn { account_id: String, amount: f64, description: String },
    OverdraftLimitSet { account_id: String, limit: f64 },
    AccountClosed { account_id: String },
}

// ─── Commands ────────────────────────────────────────────────────────────────
// Commands express intent. They can be rejected if business rules are violated.
// The command handler validates against current state before emitting events.

#[derive(Debug, Clone)]
enum AccountCommand {
    OpenAccount { account_id: String, initial_deposit: f64, owner: String },
    Deposit { account_id: String, amount: f64, description: String },
    Withdraw { account_id: String, amount: f64, description: String },
    SetOverdraftLimit { account_id: String, limit: f64 },
    CloseAccount { account_id: String },
}

// ─── Domain Errors ───────────────────────────────────────────────────────────

#[derive(Debug, Clone, PartialEq)]
enum DomainError {
    AccountAlreadyOpen(String),
    AccountNotFound(String),
    AccountClosed(String),
    InsufficientFunds { balance: f64, requested: f64, overdraft_limit: f64 },
    InvalidAmount(f64),
    CannotDepositToClosedAccount(String),
    CannotWithdrawFromClosedAccount(String),
}

impl std::fmt::Display for DomainError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            DomainError::AccountAlreadyOpen(id) => write!(f, "Account {} is already open", id),
            DomainError::AccountNotFound(id) => write!(f, "Account {} not found", id),
            DomainError::AccountClosed(id) => write!(f, "Account {} is closed", id),
            DomainError::InsufficientFunds { balance, requested, overdraft_limit } => {
                write!(f, "Insufficient funds: balance={}, requested={}, overdraft_limit={}", balance, requested, overdraft_limit)
            }
            DomainError::InvalidAmount(amt) => write!(f, "Invalid amount: {}", amt),
            DomainError::CannotDepositToClosedAccount(id) => write!(f, "Cannot deposit to closed account {}", id),
            DomainError::CannotWithdrawFromClosedAccount(id) => write!(f, "Cannot withdraw from closed account {}", id),
        }
    }
}

impl std::error::Error for DomainError {}

// ─── Aggregate State ─────────────────────────────────────────────────────────
// The aggregate rebuilds its state by applying events in order.
// This is the core of event sourcing: state = fold(apply, events).

#[derive(Debug, Clone)]
struct AccountState {
    account_id: String,
    owner: String,
    balance: f64,
    overdraft_limit: f64,
    is_open: bool,
}

impl AccountState {
    fn default_for(account_id: &str) -> Self {
        AccountState {
            account_id: account_id.to_string(),
            owner: String::new(),
            balance: 0.0,
            overdraft_limit: 0.0,
            is_open: false,
        }
    }

    fn apply(&mut self, event: &AccountEvent) {
        match event {
            AccountEvent::AccountOpened { account_id, initial_deposit, owner } => {
                self.account_id = account_id.clone();
                self.owner = owner.clone();
                self.balance = *initial_deposit;
                self.is_open = true;
            }
            AccountEvent::MoneyDeposited { amount, .. } => {
                self.balance += amount;
            }
            AccountEvent::MoneyWithdrawn { amount, .. } => {
                self.balance -= amount;
            }
            AccountEvent::OverdraftLimitSet { limit, .. } => {
                self.overdraft_limit = *limit;
            }
            AccountEvent::AccountClosed { .. } => {
                self.is_open = false;
            }
        }
    }

    fn from_events(account_id: &str, events: &[AccountEvent]) -> Self {
        let mut state = Self::default_for(account_id);
        for event in events {
            state.apply(event);
        }
        state
    }
}

// ─── Event Store ─────────────────────────────────────────────────────────────
// The event store is append-only. Events are never modified or deleted.
// In production, this would be EventStoreDB, Kafka, or a database with an
// append-only table.

#[derive(Debug, Clone)]
struct EventStore {
    streams: HashMap<String, Vec<AccountEvent>>,
}

impl EventStore {
    fn new() -> Self {
        EventStore {
            streams: HashMap::new(),
        }
    }

    fn append(&mut self, account_id: &str, events: Vec<AccountEvent>) {
        let stream = self.streams.entry(account_id.to_string()).or_insert_with(Vec::new);
        for event in events {
            stream.push(event);
        }
    }

    fn load(&self, account_id: &str) -> Vec<AccountEvent> {
        self.streams
            .get(account_id)
            .cloned()
            .unwrap_or_default()
    }

    fn load_from_version(&self, account_id: &str, from_version: usize) -> Vec<AccountEvent> {
        self.streams
            .get(account_id)
            .map(|stream| stream[from_version..].to_vec())
            .unwrap_or_default()
    }

    fn version(&self, account_id: &str) -> usize {
        self.streams
            .get(account_id)
            .map(|stream| stream.len())
            .unwrap_or(0)
    }
}

// ─── Command Handler ─────────────────────────────────────────────────────────
// The command handler is the write side of CQRS. It:
// 1. Loads the aggregate by replaying events from the store
// 2. Validates the command against current state
// 3. Emits new events if the command is valid
// It never returns data — only events or errors.

fn handle_command(
    store: &mut EventStore,
    command: &AccountCommand,
) -> Result<Vec<AccountEvent>, DomainError> {
    match command {
        AccountCommand::OpenAccount { account_id, initial_deposit, owner } => {
            if *initial_deposit < 0.0 {
                return Err(DomainError::InvalidAmount(*initial_deposit));
            }
            let existing = store.load(account_id);
            if !existing.is_empty() {
                return Err(DomainError::AccountAlreadyOpen(account_id.clone()));
            }
            Ok(vec![AccountEvent::AccountOpened {
                account_id: account_id.clone(),
                initial_deposit: *initial_deposit,
                owner: owner.clone(),
            }])
        }

        AccountCommand::Deposit { account_id, amount, description } => {
            if *amount <= 0.0 {
                return Err(DomainError::InvalidAmount(*amount));
            }
            let events = store.load(account_id);
            if events.is_empty() {
                return Err(DomainError::AccountNotFound(account_id.clone()));
            }
            let state = AccountState::from_events(account_id, &events);
            if !state.is_open {
                return Err(DomainError::CannotDepositToClosedAccount(account_id.clone()));
            }
            Ok(vec![AccountEvent::MoneyDeposited {
                account_id: account_id.clone(),
                amount: *amount,
                description: description.clone(),
            }])
        }

        AccountCommand::Withdraw { account_id, amount, description } => {
            if *amount <= 0.0 {
                return Err(DomainError::InvalidAmount(*amount));
            }
            let events = store.load(account_id);
            if events.is_empty() {
                return Err(DomainError::AccountNotFound(account_id.clone()));
            }
            let state = AccountState::from_events(account_id, &events);
            if !state.is_open {
                return Err(DomainError::CannotWithdrawFromClosedAccount(account_id.clone()));
            }
            let available = state.balance + state.overdraft_limit;
            if available < *amount {
                return Err(DomainError::InsufficientFunds {
                    balance: state.balance,
                    requested: *amount,
                    overdraft_limit: state.overdraft_limit,
                });
            }
            Ok(vec![AccountEvent::MoneyWithdrawn {
                account_id: account_id.clone(),
                amount: *amount,
                description: description.clone(),
            }])
        }

        AccountCommand::SetOverdraftLimit { account_id, limit } => {
            if *limit < 0.0 {
                return Err(DomainError::InvalidAmount(*limit));
            }
            let events = store.load(account_id);
            if events.is_empty() {
                return Err(DomainError::AccountNotFound(account_id.clone()));
            }
            let state = AccountState::from_events(account_id, &events);
            if !state.is_open {
                return Err(DomainError::AccountClosed(account_id.clone()));
            }
            Ok(vec![AccountEvent::OverdraftLimitSet {
                account_id: account_id.clone(),
                limit: *limit,
            }])
        }

        AccountCommand::CloseAccount { account_id } => {
            let events = store.load(account_id);
            if events.is_empty() {
                return Err(DomainError::AccountNotFound(account_id.clone()));
            }
            let state = AccountState::from_events(account_id, &events);
            if !state.is_open {
                return Err(DomainError::AccountClosed(account_id.clone()));
            }
            Ok(vec![AccountEvent::AccountClosed {
                account_id: account_id.clone(),
            }])
        }
    }
}

// ─── Projections (Read Model) ────────────────────────────────────────────────
// Projections consume events and build read-optimized views. This is the
// query side of CQRS. Multiple projections can consume the same event stream.

#[derive(Debug, Clone)]
struct BalanceView {
    account_id: String,
    balance: f64,
    overdraft_limit: f64,
    available: f64,
    is_open: bool,
}

struct BalanceProjection;

impl BalanceProjection {
    fn project(account_id: &str, events: &[AccountEvent]) -> BalanceView {
        let state = AccountState::from_events(account_id, events);
        BalanceView {
            account_id: state.account_id.clone(),
            balance: state.balance,
            overdraft_limit: state.overdraft_limit,
            available: state.balance + state.overdraft_limit,
            is_open: state.is_open,
        }
    }
}

#[derive(Debug, Clone)]
struct TransactionEntry {
    event_type: String,
    amount: f64,
    description: String,
}

struct HistoryProjection;

impl HistoryProjection {
    fn project(events: &[AccountEvent]) -> Vec<TransactionEntry> {
        events.iter().map(|event| match event {
            AccountEvent::AccountOpened { initial_deposit, owner, .. } => TransactionEntry {
                event_type: "OPENED".to_string(),
                amount: *initial_deposit,
                description: format!("Account opened by {}", owner),
            },
            AccountEvent::MoneyDeposited { amount, description, .. } => TransactionEntry {
                event_type: "DEPOSIT".to_string(),
                amount: *amount,
                description: description.clone(),
            },
            AccountEvent::MoneyWithdrawn { amount, description, .. } => TransactionEntry {
                event_type: "WITHDRAWAL".to_string(),
                amount: -*amount,
                description: description.clone(),
            },
            AccountEvent::OverdraftLimitSet { limit, .. } => TransactionEntry {
                event_type: "OVERDRAFT_SET".to_string(),
                amount: *limit,
                description: format!("Overdraft limit set to {}", limit),
            },
            AccountEvent::AccountClosed { .. } => TransactionEntry {
                event_type: "CLOSED".to_string(),
                amount: 0.0,
                description: "Account closed".to_string(),
            },
        }).collect()
    }
}

// ─── Snapshots ───────────────────────────────────────────────────────────────
// Snapshots save the aggregate state at a point in time to avoid replaying
// all events on every command. Load the snapshot, then replay only events
// that came after.

const SNAPSHOT_INTERVAL: usize = 3;

#[derive(Debug, Clone)]
struct Snapshot {
    state: AccountState,
    version: usize,
}

struct SnapshotStore {
    snapshots: HashMap<String, Snapshot>,
}

impl SnapshotStore {
    fn new() -> Self {
        SnapshotStore { snapshots: HashMap::new() }
    }

    fn save(&mut self, account_id: &str, state: AccountState, version: usize) {
        self.snapshots.insert(account_id.to_string(), Snapshot { state, version });
    }

    fn load(&self, account_id: &str) -> Option<&Snapshot> {
        self.snapshots.get(account_id)
    }
}

fn load_with_snapshot(
    store: &EventStore,
    snapshot_store: &SnapshotStore,
    account_id: &str,
) -> AccountState {
    if let Some(snapshot) = snapshot_store.load(account_id) {
        let remaining = store.load_from_version(account_id, snapshot.version);
        let mut state = snapshot.state.clone();
        for event in &remaining {
            state.apply(event);
        }
        state
    } else {
        let events = store.load(account_id);
        AccountState::from_events(account_id, &events)
    }
}

fn maybe_save_snapshot(
    snapshot_store: &mut SnapshotStore,
    store: &EventStore,
    account_id: &str,
) {
    let version = store.version(account_id);
    if version > 0 && version % SNAPSHOT_INTERVAL == 0 {
        let events = store.load(account_id);
        let state = AccountState::from_events(account_id, &events);
        snapshot_store.save(account_id, state, version);
    }
}

// ─── Query Side ───────────────────────────────────────────────────────────────
// The query model reads from projections only. It never touches the event store
// directly for queries — that's the whole point of CQRS. Commands go through
// the command handler; queries go through projections.

struct QueryService<'a> {
    store: &'a EventStore,
}

impl<'a> QueryService<'a> {
    fn new(store: &'a EventStore) -> Self {
        QueryService { store }
    }

    fn get_balance(&self, account_id: &str) -> Option<BalanceView> {
        let events = self.store.load(account_id);
        if events.is_empty() {
            return None;
        }
        Some(BalanceProjection::project(account_id, &events))
    }

    fn get_history(&self, account_id: &str) -> Option<Vec<TransactionEntry>> {
        let events = self.store.load(account_id);
        if events.is_empty() {
            return None;
        }
        Some(HistoryProjection::project(&events))
    }
}

// ─── Demo ─────────────────────────────────────────────────────────────────────

fn main() {
    let mut store = EventStore::new();
    let mut snapshots = SnapshotStore::new();
    let account_id = "acc-001";

    println!("=== CQRS + Event Sourcing: Bank Account Demo ===\n");

    // Open account
    let events = handle_command(&mut store, &AccountCommand::OpenAccount {
        account_id: account_id.to_string(),
        initial_deposit: 100.0,
        owner: "Alice".to_string(),
    }).unwrap();
    store.append(account_id, events);
    maybe_save_snapshot(&mut snapshots, &store, account_id);

    // Deposit
    let events = handle_command(&mut store, &AccountCommand::Deposit {
        account_id: account_id.to_string(),
        amount: 50.0,
        description: "Salary".to_string(),
    }).unwrap();
    store.append(account_id, events);
    maybe_save_snapshot(&mut snapshots, &store, account_id);

    // Set overdraft limit
    let events = handle_command(&mut store, &AccountCommand::SetOverdraftLimit {
        account_id: account_id.to_string(),
        limit: 200.0,
    }).unwrap();
    store.append(account_id, events);
    maybe_save_snapshot(&mut snapshots, &store, account_id);

    // Withdraw within balance
    let events = handle_command(&mut store, &AccountCommand::Withdraw {
        account_id: account_id.to_string(),
        amount: 80.0,
        description: "Groceries".to_string(),
    }).unwrap();
    store.append(account_id, events);
    maybe_save_snapshot(&mut snapshots, &store, account_id);

    // Withdraw into overdraft
    let events = handle_command(&mut store, &AccountCommand::Withdraw {
        account_id: account_id.to_string(),
        amount: 200.0,
        description: "Emergency repair".to_string(),
    }).unwrap();
    store.append(account_id, events);
    maybe_save_snapshot(&mut snapshots, &store, account_id);

    // Try to exceed overdraft — should fail
    let result = handle_command(&mut store, &AccountCommand::Withdraw {
        account_id: account_id.to_string(),
        amount: 100.0,
        description: "Should fail".to_string(),
    });
    println!("Withdrawal exceeding overdraft: {:?}", result.unwrap_err());

    // Query: current balance via projection
    let query = QueryService::new(&store);
    let balance_view = query.get_balance(account_id).unwrap();
    println!("\n--- Balance View (Read Model) ---");
    println!("Account: {}", balance_view.account_id);
    println!("Balance: ${:.2}", balance_view.balance);
    println!("Overdraft Limit: ${:.2}", balance_view.overdraft_limit);
    println!("Available: ${:.2}", balance_view.available);
    println!("Is Open: {}", balance_view.is_open);

    // Query: transaction history via projection
    let history = query.get_history(account_id).unwrap();
    println!("\n--- Transaction History (Read Model) ---");
    for entry in &history {
        println!("{:<15} {:>8.2}  {}", entry.event_type, entry.amount, entry.description);
    }

    // Verify snapshot replay matches full replay
    let full_state = {
        let events = store.load(account_id);
        AccountState::from_events(account_id, &events)
    };
    let snapshot_state = load_with_snapshot(&store, &snapshots, account_id);
    println!("\n--- Snapshot Verification ---");
    println!("Full replay balance: ${:.2}", full_state.balance);
    println!("Snapshot replay balance: ${:.2}", snapshot_state.balance);
    assert!((full_state.balance - snapshot_state.balance).abs() < f64::EPSILON);
    println!("Snapshot replay matches full replay: OK");

    // Show the event store
    println!("\n--- Event Store (Immutable Log) ---");
    for (i, event) in store.load(account_id).iter().enumerate() {
        println!("v{}: {:?}", i + 1, event);
    }

    // Close account
    let events = handle_command(&mut store, &AccountCommand::CloseAccount {
        account_id: account_id.to_string(),
    }).unwrap();
    store.append(account_id, events);

    // Try deposit on closed account — should fail
    let result = handle_command(&mut store, &AccountCommand::Deposit {
        account_id: account_id.to_string(),
        amount: 10.0,
        description: "Should fail".to_string(),
    });
    println!("\nDeposit to closed account: {}", result.unwrap_err());

    println!("\n=== All business rules enforced. CQRS + ES demo complete. ===");
}